manifest.go

```diff
@@ -614,6 +559,211 @@ func decodeManifests[I interface {
+			contentStr)
+	}
+	if content != file.ManifestContent() {
+		return nil, fmt.Errorf("manifest file's 'content' metadata indicates %q, but entry from manifest list indicates %q",
+			content.String(), file.ManifestContent().String())
+	}
+
+	isFallback := false
+	if formatVersion == 1 {
+		for _, f := range sc.(*avro.RecordSchema).Fields() {
+			if f.Name() == "snapshot_id" {
+				if f.Type().Type() != avro.Union {
+					isFallback = true
+				}
+
+				break
+			}
+		}
+	}
+	fieldNameToID, fieldIDToType := getFieldIDMap(sc)
+
+	return &ManifestReader{
+		dec:           dec,
+		file:          file,
+		formatVersion: formatVersion,
+		isFallback:    isFallback,
+		content:       content,
+		fieldNameToID: fieldNameToID,
+		fieldIDToType: fieldIDToType,
+	}, nil
+}
+
+// Version returns the file's format version.
+func (c *ManifestReader) Version() int {
+	return c.formatVersion
+}
+
+// ManifestContent returns the type of content in the manifest file.
+func (c *ManifestReader) ManifestContent() ManifestContent {
+	return c.content
+}
+
+// SchemaID returns the schema ID encoded in the avro file's metadata.
+func (c *ManifestReader) SchemaID() (int, error) {
+	id, err := strconv.Atoi(string(c.dec.Metadata()["schema-id"]))
+	if err != nil {
+		return 0, fmt.Errorf("manifest file's 'schema-id' metadata is invalid: %w", err)
+	}
+
+	return id, nil
+}
+
+// Schema returns the schema encoded in the avro file's metadata.
+func (c *ManifestReader) Schema() (*Schema, error) {
+	if !c.schemaLoaded {
+		schemaID, err := c.SchemaID()
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(c.dec.Metadata()["schema"], &c.schema); err != nil {
+			return nil, fmt.Errorf("manifest file's 'schema' metadata is invalid: %w", err)
+		}
+		c.schema.ID = schemaID
+		c.schemaLoaded = true
+	}
+
+	return &c.schema, nil
+}
+
+// PartitionSpecID returns the partition spec ID encoded in the avro file's metadata.
+func (c *ManifestReader) PartitionSpecID() (int, error) {
+	id, err := strconv.Atoi(string(c.dec.Metadata()["partition-spec-id"]))
+	if err != nil {
+		return 0, fmt.Errorf("manifest file's 'partition-spec-id' metadata is invalid: %w", err)
+	}
+	if id != int(c.file.PartitionSpecID()) {
+		return 0, fmt.Errorf("manifest file's 'partition-spec-id' metadata indicates %d, but entry from manifest list indicates %d",
+			id, c.file.PartitionSpecID())
+	}
+
+	return id, nil
+}
+
+// PartitionSpec returns the partition spec encoded in the avro file's metadata.
+func (c *ManifestReader) PartitionSpec() (*PartitionSpec, error) {
+	if !c.partitionSpecLoaded {
+		partitionSpecID, err := c.PartitionSpecID()
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(c.dec.Metadata()["partition-spec"], &c.partitionSpec.fields); err != nil {
+			return nil, fmt.Errorf("manifest file's 'partition-spec' metadata is invalid: %w", err)
+		}
+		c.partitionSpec.id = partitionSpecID
+		c.partitionSpec.initialize()
+		c.partitionSpecLoaded = true
+	}
```

Comment:
same question as above, do we care or worry about making sure to avoid a race condition here?